### ISSUE TITLE
New command to modify filters at runtime

### ISSUE THREAD

**Issue by yottapulsky** [namhyung/uftrace#1330]
PR #1269 introduces a new uftrace command, able to modify the filters of an already running uftrace instance. Feedback is needed about some design features.

The PR comes with a handful of new features, such as modifying filters and triggers, as well as time and depth options. It can be reviewed independently from PRs than rely on it (namely #1274).

# Motivation

Some use cases require instrumenting long running processes, whose state and environment may not be predictable, and which cannot be restarted. The use of static triggers is not fexible enough because the parameters are to be adapted for specific situations. Being able to interact with uftrace at runtime would allow for more flexibility.

Furthermore, this paves the way for fully dynamic instrumentation, as it provides a way to tell uftrace which function to (un)patch (see #1274). The first step would be to implement the client structure, which will provide the entry point for dynamic patching at runtime.

# Architecture

The proposed structure is a client/server architecture. See picture below.

A daemon thread is ran at mcount startup. It opens a socket at `/var/run/user/$UID/uftrace/$PID.socket`, and waits for incoming commands.
The new command then forwards its options to the daemon through the socket, locating it via its PID.

![uftrace_client](https://user-images.githubusercontent.com/33483519/131741799-2ce9ee53-ff95-470e-9786-1dc506957c08.png)

The daemon then calls setup functions to alter the filters, and modifies the appropriate global variables.

## Discussion

The following points require feedback:

- [ ] the communication mechanism: integers sent through a UNIX socket (RPC would be overkill)
- [ ] the location of the socket `/var/run/user/$UID/uftrace/$PID.socket`, where $PID is the PID of the process hosting daemon

# Integration

The daemon-powered mcount library builds as a separate unit: `libmcount-daemon.so`. It is used when the `--daemon` option is provided.

The new command is currently named `uftrace-client` in the PR, referring to its role in the client/server architecture. On the other hand, the `uftrace-apply` would be more consistent with the actual behavior (i.e. applying options) and with the naming of other commands (e.g. record). I actually suggest using `apply`.

## Discussion

The following points require feedback:

- [ ] separate build unit for the daemon: `libmcount-daemon.so`
- [ ] name of the command: `client` v `apply`

**Comment by namhyung** [comment]
@yottapulsky thanks for your detailed explanation.  It helps me to understand the issue you want to solve.  So do you want to run uftrace continuously and also want to change filters/triggers?

I've just got a different idea.  As I said, eventually I want to have `uftrace record -p <PID>`.  I'd like to not have `uftrace client` or (`apply`) if possible.  I think we can make `uftrace record --daemon` to preload the libmcount-daemon.so just to setup the socket.  I think the uftrace record process can exit at this point.  Then we can use `uftracd record -p <PID>` to connect to the target process with the socket.  The libmcount-daemon.so will initialize filters/triggers according to the option in the record process and send results to it.  When the record process finishes, libmcount-daemon.so will clean up all the states and go back to the initail state so that it can accept another request.  Maybe we need a (file) lock to protect it from concurrent accesses.  Later we can inject the libmcount-daemon.so runtime (i.e. at the time of `uftrace record -p <PID>` and it the socket is not found) to eliminate the initial `uftrace record --daemon`.

Would it work in your use case?  I think you can (re)start `uftrace record -p <PID>` with different options if you want to change it runtime.

**Comment by yottapulsky** [comment]
Yes, the point was having uftrace continuously tracing its target, and in the meantime change the filters/triggers (and eventually (un)patch functions) in the target.

Having the `-p <PID>` option in the `record` command seems intuitive, as one can choose to trace either a binary or a process. Eventually the record command will be able to load `libmcount-*.so` at runtime, but for the moment, let's assume `libmcount-daemon.so` is preloaded at startup as you suggest.

Is `libmcount` supposed to trace the target before the `record` process starts? And after it finishes, while waiting for another request? Because it won't be able to write the events to disk.

Suppose `libmcount` continues tracing. The problem I see with only having the `record` command is that you have to restart it every time you want to apply new options. And in the meantime `libmcount` would go back to its initial state, altering the filters.
The `client` command is meant to be called multiple times, and to make its modifications persistent between the calls.
Is it necessary to have the `libmcount` state reset when `record` finishes? If not, a solution would be to make the changes persistent. But repetitive calls to `uftrace record` may not be desirable, as they would be used to _apply_ options rather than to start recording.

On the other hand, if tracing stops when `record` finishes, we would lose flexibility as we could miss events while the `restart` command is restarted.

**Comment by namhyung** [comment]
I didn't expect you want to trace it all the time because of the data size.  At some point you would need to disable tracing unless the data traffic is very low.  So I think it's ok to pause tracing by stopping and restarting record with new option.

**Comment by yottapulsky** [comment]
I agree that in most cases it should be OK to stop tracing.

We actually have use cases where we really need to keep tracing, and where the data traffic is indeed low. For example, we'd like to change the filters the way triggers do, but with an external script reading the trace output and using it to make decisions. We can discuss about this particular use case in another issue. It's an example of a case where we would need continuous tracing.

Do you really think we cannot have a `client`/`apply` command? It corresponds to a new and distinct behavior, and having its logic separated from the `record` command is more intuitive IMO.
If we are to only keep the `record` command, it would be nice to have the ability to modify options while tracing continuously, but I'm not sure it is possible without complicating `record` too much.

**Comment by namhyung** [comment]
How did you use the external script?  By uftrace?

If you really need the continuous tracing, yeah maybe we should add the client/apply command.  I'm just thinking about the long term maintenance burden it'd bring, as it breaks the assumption in the stability of the shared data structure like filters.

**Comment by yottapulsky** [comment]
This feature is only an idea, we don't have any script yet. It could be used with the `script` command, but we need to execute it while the trace is being written. Currently this is not possible.

Yes, we'd really like to have it. What is problematic exactly? Do you think there is a way to manipulate the shared structures so the stability isn't impacted too much?

**Comment by namhyung** [comment]
Note that `uftrace record -S <script>` can run a script during record.

I designed it not to modify the global data structures (like filters) once it's initialized mostly for the performance reason.  But we can do that for the daemon case in a separate libmcount-daemon.so.  I think we can share the same code in the libmcount and add some conditional codes to do the necessary jobs for the daemon case only.

**Comment by honggyukim** [comment]
Sorry for jumping into the discussion late. The expected execution model of @namhyung looks as follows.
```
$ uftrace record --daemon target_service
target_service(pid 12345) is ready to be traced, but stopped for now.
Please run 'uftrace record -p 12345 <options>' to start execution and tracing.

$ uftrace record -p 12345 -D 5

$ uftrace record -p 12345 -F func2

(... target_service execution is finished ...)

$ uftrace replay
```
But in this case, it seems that we will have two separate `uftrace.data` directories for each `uftrace record`.

> When the record process finishes, libmcount-daemon.so will clean up all the states and go back to the initail state so that it can accept another request.

When does the record process finishes? The issue I think is we don't have a way to detach `uftrace` until the program finishes. If so, we can't change the filter in the middle of execution. We may need another option something like `--detach-after=TIME` or have to change the behavior of `@finish` trigger action.

**Comment by honggyukim** [comment]
If we need to support continuous tracing, then I also think there should be another way such as `client` or `apply` command because `uftrace record` is expected to drop `uftrace.data` directory for each run.

But the issue is the implementation and maintenance in the long term as @namhyung mentioned. We should lock/unlock for each access to `mcount_triggers`, `mcount_depth`, `mcount_threshold` and many other related data structures. These locking can make performance overhead in general cases so should only be used in `libmcount-daemon.so`.

**Comment by yottapulsky** [comment]
Thank you for the explanations.

From my last changes in #1269, the daemon was only started in `libmcount-daemon.so`, which is not selected by default. So locks are necessary only is this library.

I'll be working on making `record` able to interact with the daemon too, and to detach from the target.

**Comment by namhyung** [comment]
I'm afraid that attaching/detaching to an existing process (with or without an existing uftrace daemon) is not simple.

**Comment by yottapulsky** [comment]
I agree. The first step would be to attach to an existing daemon I think, but for this we need to setup libmcount (e.g. shared memory buffers) asynchronously. I'll be putting my efforts on dynamic patching first, as it can use my current client.

Regarding the detach part, @honggyukim suggested using `--detach-after=TIME`. Maybe we can use a `--detach` option in the client too, that can order libmcount to make the record process to detach.

**Comment by namhyung** [comment]
If it can support attach/detach, I think the most intuitive way to detach is to catch Ctrl-C in uftrace record and send a command to the daemon.

**Comment by honggyukim** [comment]
> I think the most intuitive way to detach is to catch Ctrl-C in uftrace record and send a command to the daemon.

I think it's also a good way to detach, but not sure if it can reliably work as expected. I personally think we can also have `--detach-after=TIME` for different purposes. This can be done in a separate work.
